Fix production mode inconsistent CSS rendering when dynamic property values are the same but property names are different

bcole808 opened the pull request:
This PR fixes https://github.com/styled-components/styled-components/issues/4113

This problem occurs when dynamic interpolations in different places can return the same value. 

An example of how this problem can happen:

```ts
const MyComponent = styled.div<{ $variant: 'text' | 'background' }>`
      color: ${props => props.$variant == 'text' && 'red'};
      background-color: ${props => props.$variant == 'background' && 'red'};
`;
```

In the example above, there are two interpolation functions. Each returns the exact same value, but appears next to a different CSS property name. 

In production builds, the logic which determines if a new class name needs to be created fails to check for this situation and outputs the same CSS for both `<MyComponent $variant="text" />` and `<MyComponent $variant="background" />`.

In non-production builds, this problem does not occur.

guillaumebrunerie commented:
Nice! Looking at the log, the latest commit that changed this line was https://github.com/styled-components/styled-components/commit/d5fe2495b94a037234147bba6febfd8473dd86ac which removed the `+ i` from the two lines `dynamicHash = phash(dynamicHash, partString + i);`. Maybe that would fix it as well? This `+ i` was added in https://github.com/styled-components/styled-components/commit/fff5c1e4ab30ddf131dd6ff780f6a52ac480287d to fix an issue very similar to my bug report.

bcole808 commented:
@guillaumebrunerie Good observation! 

After looking at those commits more closely, I think https://github.com/styled-components/styled-components/commit/d5fe2495b94a037234147bba6febfd8473dd86ac is the commit that introduced this bug. The `+ 1` for parts that are NOT a string was important for this use case. 

This bug happens when there is something dynamic (i.e. a function) in the CSS which results in the same hash but could change positions in the list of interpolations.  Removing the `if (process.env.NODE_ENV !== 'production')` conditional fixes the problem by accounting for parts of the CSS which are static strings, which accounts for when a dynamic interpolation changes position in the list of interpolations. 

### What about performance of a production build?

By removing `if (process.env.NODE_ENV !== 'production') `, **more** hashing happens in production builds. 

I ran the benchmarks briefly on my machine and observed a speed of **4.36 on main** compared to **4.43 on this PR**. This represents a decrease in production performance of about 1.6% (If someone else can also run the benchmarks that would be great)

### Is there a way we can solve this bug without impacting production performance?

Yes, A different way could be to put `+ i` back into the other check for when `partRule` is not a string.  The commit message on https://github.com/styled-components/styled-components/commit/fff5c1e4ab30ddf131dd6ff780f6a52ac480287d explains: 

> When the same interpolations swap places they still produce the same hash which can sometimes lead to the same hash.

```ts
else if (partRule) {
    const partString = joinStringArray(
      flatten(partRule, executionContext, styleSheet, stylis) as string[]
    );
    dynamicHash = phash(dynamicHash, partString + i); // <---- Add + i here
    css += partString;
}
```

This would **also** solve the problem. 

### So what are the options?

**Option 1: Remove `if (process.env.NODE_ENV !== 'production')`**

* This makes the behavior consistent between production and development modes at the cost of impacting production performance slightly. 

**Option 2: Add `+ i` back into the hash for non-strings**

* This fixes the issue without affecting performance, but arguably is harder to understand and maintain because it isn't obvious why `+ i` is necessary. 

@probablyup What do you think? What is the right balance here between production performance and understandability?

probablyup commented on packages/styled-components/src/test/basic.test.tsx:
```diff
@@ -66,6 +66,23 @@ describe('basic', () => {
     `);
   });
 
+  it('should inject two different styles if the same compnoent is mounted with different props and css', () => {
```
```suggestion
  it('should inject two different styles if the same component is mounted with different props and css', () => {
```

probablyup commented:
Performance is more important, let's go with option 2. Code comments appreciated :)

bcole808 commented:
Thanks @probablyup, I've switched to option 2 and added a comment to explain it. The added unit test should also catch this regression in the future as well. 

PS: I'd love to be added as a contributor so I can push branches directly to the repo, and also help review PRs! It looks like there used to be a mechanism to auto-invite contributors, but it's no longer working. [I posted over here about it as well.](https://github.com/orgs/styled-components/discussions/4142)

probablyup commented:
One other option would be to derive from domElements.ts. That's actually
the list that s-c utilizes so would be even smaller than intrinsic elements
since not every html element can be styled

On Mon, Sep 11, 2023 at 12:49 PM Ben Cole ***@***.***> wrote:

> Thanks @probablyup <https://github.com/probablyup>, I've switched to
> option 2 and added a comment to explain it. The added unit test should also
> catch this regression in the future as well.
>
> PS: I'd love to be added as a contributor so I can push branches directly
> to the repo, and also help review PRs! It looks like there used to be a
> mechanism to auto-invite contributors, but it's no longer working. I
> posted over here about it as well.
> <https://github.com/orgs/styled-components/discussions/4142>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/styled-components/styled-components/pull/4132#issuecomment-1714245992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAELFVVI72AAO6NTY3XRTADXZ46H5ANCNFSM6AAAAAA3V4PUDQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bcole808 commented:
Not sure I follow what you're saying? I don't think this PR has anything to do with specific html elements, right?

probablyup commented:
Sorry this was meant for another PR

probablyup reviewed (approved):
Awesome, thanks for looking into this.
